fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java

```diff
@@ -0,0 +1,203 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.collect.Lists;
+import com.starrocks.common.Pair;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.Utils;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalOlapScanOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalUnionOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.scalar.FilterSelectivityEvaluator;
+import com.starrocks.sql.optimizer.rewrite.scalar.FilterSelectivityEvaluator.ColumnFilter;
+import com.starrocks.sql.optimizer.rewrite.scalar.NegateFilterShuttle;
+import com.starrocks.sql.optimizer.rule.RuleType;
+import com.starrocks.sql.optimizer.statistics.Statistics;
+import com.starrocks.sql.optimizer.statistics.StatisticsCalcUtils;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+// For sql like:
+//      select * from tbl where col1 > 100 and (col2 = 1 or col3 =2 or col4 = 4), the (col2 = 1 or col3 =2 or col4 = 4)
+// predicate can actually filter lots of rows, but our index doesn't support this predicate. We can transform it to:
+//      select * from tbl where  col1 > 100 and col2 = 1
+//      union all
+//      select * from tbl where col1 > 100 and col3 = 2 and (col2 !=1 or col2 is null)
+//      union all
+//      select * from tbl where col1 > 100 and col4 = 4 and (col2 !=1 or col2 is null) and (col3 != 2 or col3 is null)
+// Every scanNode has an equivalent predicate can use index to filter lots of rows.
+public class SplitScanORToUnionRule extends TransformationRule {
+    private static final Logger LOG = LogManager.getLogger(SplitScanORToUnionRule.class);
+    private static final double SELECTIVITY_THRESHOLD = 0.035;
```

Comment:
why choose a magic value 0.035?  why not use sessiable variable to control this?

Reply:
If a predicate can continue to filter out data close to 29/30 after partition pruning, I think there is no need to do this optimization. Considering that this rule has already added two session variables, I don’t want to add the third one.

Reply:
It seems this parameter are more meaningful than other two parameters. because selectivity of predicates are accurate,  in a real product environment, this parameter maybe need to be adjusted.